README.md
# js-excel-template

[![Dependency Status](https://david-dm.org/plantain-00/js-excel-template.svg)](https://david-dm.org/plantain-00/js-excel-template)
[![devDependency Status](https://david-dm.org/plantain-00/js-excel-template/dev-status.svg)](https://david-dm.org/plantain-00/js-excel-template#info=devDependencies)
[![Build Status: Windows](https://ci.appveyor.com/api/projects/status/github/plantain-00/js-excel-template?branch=master&svg=true)](https://ci.appveyor.com/project/plantain-00/js-excel-template/branch/master)
![Github CI](https://github.com/plantain-00/js-excel-template/workflows/Github%20CI/badge.svg)
[![npm version](https://badge.fury.io/js/js-excel-template.svg)](https://badge.fury.io/js/js-excel-template)
[![Downloads](https://img.shields.io/npm/dm/js-excel-template.svg)](https://www.npmjs.com/package/js-excel-template)
[![gzip size](https://img.badgesize.io/https://unpkg.com/js-excel-template?compression=gzip)](https://unpkg.com/js-excel-template)
[![type-coverage](https://img.shields.io/badge/dynamic/json.svg?label=type-coverage&prefix=%E2%89%A5&suffix=%&query=$.typeCoverage.atLeast&uri=https%3A%2F%2Fraw.githubusercontent.com%2Fplantain-00%2Fjs-excel-template%2Fmaster%2Fpackage.json)](https://github.com/plantain-00/js-excel-template)

A js excel template used in browser or nodejs environment.

## install

`npm i js-excel-template`

## features

+ Generate Excel based on Excel template
+ (for nodejs)Read excel file as template
+ (for nodejs)Write to excel file
+ (for nodejs)To a `Buffer` for downloading
+ (for browser-side js)Take `ArrayBuffer` as template, coming from HTTP downloading, browser uploading or Websocket downloading
+ (for browser-side js)To a `ArrayBuffer` or `Blob` for HTTP uploading, browser-side downloading or Websocket uploading

## example

![template](https://raw.githubusercontent.com/plantain-00/js-excel-template/master/doc/template.PNG)

```ts
// nodejs:
import JsExcelTemplate from "js-excel-template";
const excelTemplate = await JsExcelTemplate.fromFile("demo/test.xlsx");

// browser(module):
// import * as FileSaver from "file-saver";
// import JsExcelTemplate from "js-excel-template";
// const response = await fetch('./test.xlsx')
// const arrayBuffer = await response.arrayBuffer()
// const excelTemplate = await JsExcelTemplate.fromArrayBuffer(arrayBuffer)

// browser(script tag):
// <script src="file-saver/FileSaver.min.js"></script>
// <script src="exceljs/dist/exceljs.min.js"></script>
// <script src="js-excel-template/browser/js-excel-template.min.js"></script>
// <script>
//   const response = await fetch('./test.xlsx')
//   const arrayBuffer = await response.arrayBuffer()
//   const excelTemplate = await JsExcelTemplate.fromArrayBuffer(arrayBuffer)

excelTemplate.set("name", "John");
excelTemplate.set("age", 123);
excelTemplate.set("now", new Date());
excelTemplate.set("isBoy", true);
excelTemplate.set('isGirl', {
  richText: [
    {
      text: 'false',
      font: {
        strike: true,
      },
    },
  ],
})

const students = [
  { name: "Tommy", age: 12 },
  { name: "Philips", age: 13 },
  { name: "Sara", age: 14 },
];

for (let i = 1; i <= 5; i++) {
  if (i <= students.length) {
    excelTemplate.set(`name${i}`, students[i - 1].name);
    excelTemplate.set(`age${i}`, students[i - 1].age);
  } else {
    excelTemplate.set(`name${i}`, "");
    excelTemplate.set(`age${i}`, "");
  }
}
excelTemplate.set("average", students.reduce((p, c) => p + c.age, 0) / students.length);

excelTemplate.set('fields', [{ name: 'Name' }, { name: 'Age' }], { duplicateCellIfArray: true })

excelTemplate.set("students", students);

// nodejs:
await excelTemplate.saveAs("spec/out.xlsx");

// browser(module):
// const blob = await excelTemplate.toBlob()
// FileSaver.saveAs(blob, "test.xlsx");

// browser(script tag):
//   const blob = await excelTemplate.toBlob()
//   FileSaver.saveAs(blob, "test.xlsx");
// </script>
```

![out](https://raw.githubusercontent.com/plantain-00/js-excel-template/master/doc/out.PNG)

## change logs

```ts
// v3
const excelTemplate = await JsExcelTemplate.fromFile("demo/test.xlsx");
const excelTemplate = await JsExcelTemplate.fromArrayBuffer(arrayBuffer);
<script src="exceljs/dist/exceljs.min.js"></script>
await excelTemplate.saveAs("spec/out.xlsx");
const blob = await excelTemplate.toBlob()

// v2
const excelTemplate = JsExcelTemplate.fromFile("demo/test.xlsx");
const excelTemplate = JsExcelTemplate.fromArrayBuffer(arrayBuffer);
<script src="xlsx/dist/xlsx.full.min.js"></script>
excelTemplate.saveAs("spec/out.xlsx");
const blob = excelTemplate.toBlob()
```

```ts
// v2
import JsExcelTemplate from "js-excel-template/nodejs/nodejs";
import JsExcelTemplate from "js-excel-template/browser/browser";

// v1
import { JsExcelTemplate } from "js-excel-template/nodejs";
import { JsExcelTemplate } from "js-excel-template/browser";
```


CHANGELOG.md
# Change Log

## [3.0.2](https://github.com/plantain-00/js-excel-template/compare/v3.0.1...v3.0.2) (2022-12-20)
  
* fix: path error in package json ([0232799](https://github.com/plantain-00/js-excel-template/commit/023279972d39f43e22ccdba5b0851524b50e1a5f))
* chore: update dependencies ([0979f64](https://github.com/plantain-00/js-excel-template/commit/0979f64963c3f333b1508b7295d7504dd8e46cf7)) ([f8f09a9](https://github.com/plantain-00/js-excel-template/commit/f8f09a9e90c5c4f8803dcc52c42b0d7d700229ac)) ([5683770](https://github.com/plantain-00/js-excel-template/commit/568377045763ac89e6e9a60e51897d9a879aa7ff)) ([5876e01](https://github.com/plantain-00/js-excel-template/commit/5876e0194bdb8f0fdf8b5a5f756decdafe44378a)) ([ec8b60b](https://github.com/plantain-00/js-excel-template/commit/ec8b60bbe8825d40aa4ff63749663cb024721b79)) ([35e5eb7](https://github.com/plantain-00/js-excel-template/commit/35e5eb7ae1e103db334713a28b7713871ef6b208)) ([ec7aecb](https://github.com/plantain-00/js-excel-template/commit/ec7aecbadd3b13aa91e72ef3f468a53f30f0b8c5)) ([5c41af9](https://github.com/plantain-00/js-excel-template/commit/5c41af93bf61d736a82b057fef504bfaf7836dc5)) ([ccefcde](https://github.com/plantain-00/js-excel-template/commit/ccefcde189dbcc9278c0cb12a9c629d398e7faaa)) ([513e34b](https://github.com/plantain-00/js-excel-template/commit/513e34b845e1d8c7dbe7a6ba133c4d964cb3b754)) ([4cb4f6c](https://github.com/plantain-00/js-excel-template/commit/4cb4f6c978d739d33b2133a54dad2d3c1aa9815a)) ([3240cb1](https://github.com/plantain-00/js-excel-template/commit/3240cb19b1c4ffc57561e47e4a3e4fb84d2a0f3c)) ([4035643](https://github.com/plantain-00/js-excel-template/commit/403564319e38bdb4472dad57381018a09442d172)) ([29d97c5](https://github.com/plantain-00/js-excel-template/commit/29d97c58400f03ff798479112cdc9f00d6a8781f)) ([c64b90b](https://github.com/plantain-00/js-excel-template/commit/c64b90b080a7c55f0a0988e3f58720f5e6949965)) ([1234e78](https://github.com/plantain-00/js-excel-template/commit/1234e78b7806f778b80d15920bcbb48989b900cd)) ([f94f35f](https://github.com/plantain-00/js-excel-template/commit/f94f35ff1c3acd3ee074f43e167fecfa2c179057)) ([aaa468c](https://github.com/plantain-00/js-excel-template/commit/aaa468cd498fb7323b60b6446709b175ff698049)) ([e672479](https://github.com/plantain-00/js-excel-template/commit/e6724797361e9b920c840743a2ea2696a5b43e72)) ([e074c8e](https://github.com/plantain-00/js-excel-template/commit/e074c8e3b73746d0c5f4c2d1b9896a7f89f86af9))
* fix: rollup config ([4ed1cf6](https://github.com/plantain-00/js-excel-template/commit/4ed1cf6feca6a713901fc33d16a6bed2671e0cf5))

## [3.0.1](https://github.com/plantain-00/js-excel-template/compare/v3.0.0...v3.0.1) (2022-02-18)
  
* fix: more field for array value ([d69c377](https://github.com/plantain-00/js-excel-template/commit/d69c377d738f34dd71fb9a969727566b6c79a92e))

## [3.0.0](https://github.com/plantain-00/js-excel-template/compare/v2.0.4...v3.0.0) (2022-02-18)
  
* feat: use exceljs instead of xlsx ([4c10129](https://github.com/plantain-00/js-excel-template/commit/4c101298aaba3adefc2c950d42d1b1d18f9edae9))

demo/index.ts
import * as FileSaver from 'file-saver'
import JsExcelTemplate from '../dist/browser/browser'

(async () => {
  const response = await fetch('./test.xlsx')
  const arrayBuffer = await response.arrayBuffer()
  const excelTemplate = await JsExcelTemplate.fromArrayBuffer(arrayBuffer)

  excelTemplate.set('name', 'John')
  excelTemplate.set('age', 123)
  excelTemplate.set('now', new Date())
  excelTemplate.set('isBoy', true)
  excelTemplate.set('isGirl', {
    richText: [
      {
        text: 'false',
        font: {
          strike: true,
        },
      },
    ],
  })

  const students = [
    { name: 'Tommy', age: 12 },
    { name: 'Philips', age: 13 },
    { name: 'Sara', age: 14 }
  ]

  for (let i = 1; i <= 5; i++) {
    if (i <= students.length) {
      excelTemplate.set(`name${i}`, students[i - 1].name)
      excelTemplate.set(`age${i}`, students[i - 1].age)
    } else {
      excelTemplate.set(`name${i}`, '')
      excelTemplate.set(`age${i}`, '')
    }
  }
  excelTemplate.set('average', students.reduce((p, c) => p + c.age, 0) / students.length)

  excelTemplate.set('fields', [{ name: 'Name' }, { name: 'Age' }], { duplicateCellIfArray: true })

  excelTemplate.set('students', students)

  const blob = await excelTemplate.toBlob()
  FileSaver.saveAs(blob, 'test.xlsx')
})()
